dex/pair/src/safe_price_view.rs

```diff
@@ -0,0 +1,422 @@
+        // in case the searched round is bigger than the last recording round
+        // The search round is limited to the current blockchain round
+        if last_observation.recording_round < search_round {
+            let current_round = self.blockchain().get_block_round();
+            let storage_cache = StorageCache::new(self);
+            return self.compute_new_observation(
+                core::cmp::min(search_round, current_round),
+                &storage_cache.first_token_reserve,
+                &storage_cache.second_token_reserve,
+                &last_observation,
+            );
+        }
+
+        let (mut price_observation, last_search_index) = self.price_observation_by_binary_search(
+            current_index,
+            price_observations,
+            search_round,
+        );
+
+        if price_observation.recording_round > 0 {
+            return price_observation;
+        }
+
+        price_observation = self.price_observation_by_linear_interpolation(
+            price_observations,
+            search_round,
+            last_search_index,
+        );
+
+        price_observation
+    }
+
+    fn get_oldest_price_observation(
+        &self,
+        current_index: usize,
+        price_observations: &VecMapper<Self::Api, PriceObservation<Self::Api>>,
+    ) -> PriceObservation<Self::Api> {
+        require!(
+            !price_observations.is_empty(),
+            ERROR_SAFE_PRICE_OBSERVATION_DOES_NOT_EXIST
+        );
+
+        // VecMapper index starts at 1
+        let mut oldest_observation_index = 1;
+        if price_observations.len() == MAX_OBSERVATIONS {
+            oldest_observation_index = (current_index % MAX_OBSERVATIONS) + 1
+        }
+        price_observations.get(oldest_observation_index)
+    }
+
+    fn price_observation_by_binary_search(
+        &self,
+        current_index: usize,
+        price_observations: &VecMapper<Self::Api, PriceObservation<Self::Api>>,
+        search_round: Round,
+    ) -> (PriceObservation<Self::Api>, usize) {
+        let mut search_index = 1;
+        let mut left_index;
+        let mut right_index;
+        let observation_at_index_1 = price_observations.get(search_index);
+        if observation_at_index_1.recording_round <= search_round {
+            left_index = search_index;
+            right_index = current_index - 1;
+        } else {
+            left_index = current_index + 1;
+            right_index = price_observations.len();
+        }
+
+        while left_index <= right_index {
+            search_index = (left_index + right_index) / 2;
+            let price_observation = price_observations.get(search_index);
+            match price_observation.recording_round.cmp(&search_round) {
+                Ordering::Equal => return (price_observation, search_index),
+                Ordering::Less => left_index = search_index + 1,
+                Ordering::Greater => right_index = search_index - 1,
+            }
+        }
```

Comment:
I think there's a built-in way of doing this binary search.

Reply:
Not sure which function would that be. We need to get the value (which is a struct) from the vec at a specific index, and compare one variable of the the struct, with a given parameter.

Reply:
Hi! I also need to perform a binary search with the same specs you mentioned: 

> We need to get the value (which is a struct) from the vec at a specific index, and compare one variable of the the struct, with a given parameter

Please, let me know if there is a builtin function for it. Thanks!

Reply:
Slightly misremembered, as it's only for Vec: https://doc.rust-lang.org/std/vec/struct.Vec.html#method.binary_search_by